warehouse/accounts/security_policy.py

```diff
@@ -0,0 +1,238 @@
+            # because of a compromised password. So to do that, we'll need to raise a
+            # HTTPError that'll ultimately get returned to the client. This is OK to do
+            # here because we've already successfully authenticated the credentials, so
+            # it won't screw up the fall through to other authentication mechanisms
+            # (since we wouldn't have fell through to them anyways).
+            if disabled_for == DisableReason.CompromisedPassword:
+                raise _format_exc_status(
+                    BasicAuthBreachedPassword(), breach_service.failure_message_plain
+                )
+            elif disabled_for == DisableReason.AccountFrozen:
+                raise _format_exc_status(BasicAuthAccountFrozen(), "Account is frozen.")
+            else:
+                raise _format_exc_status(HTTPUnauthorized(), "Account is disabled.")
+        elif login_service.check_password(
+            user.id,
+            password,
+            tags=["mechanism:basic_auth", "method:auth", "auth_method:basic"],
+        ):
+            if breach_service.check_password(
+                password, tags=["method:auth", "auth_method:basic"]
+            ):
+                send_password_compromised_email_hibp(request, user)
+                login_service.disable_password(
+                    user.id, reason=DisableReason.CompromisedPassword
+                )
+                raise _format_exc_status(
+                    BasicAuthBreachedPassword(), breach_service.failure_message_plain
+                )
+
+            login_service.update_user(user.id, last_login=datetime.datetime.utcnow())
+            return True
+        else:
+            user.record_event(
+                tag="account:login:failure",
+                ip_address=request.remote_addr,
+                additional={"reason": "invalid_password", "auth_method": "basic"},
+            )
+            raise _format_exc_status(
+                BasicAuthFailedPassword(),
+                "Invalid or non-existent authentication information. "
+                "See {projecthelp} for more information.".format(
+                    projecthelp=request.help_url(_anchor="invalid-auth")
+                ),
+            )
+
+
+@implementer(ISecurityPolicy)
+class SessionSecurityPolicy:
+    def __init__(self):
+        self._session_helper = SessionAuthenticationHelper()
+        self._principals = []
+
+    def identity(self, request):
+        # If we're calling into this API on a request, then we want to register
+        # a callback which will ensure that the response varies based on the
+        # Cookie header.
+        request.add_response_callback(add_vary_callback("Cookie"))
+        request.authentication_method = AuthenticationMethod.SESSION
+
+        # NOTE: Why is this authenticated_userid? I guess it's already
+        # been validated by session validation?
+        userid = self._session_helper.authenticated_userid(request)
```

Comment:
Flagging for review: this is a Pyramid API so I'm assuming it's correct, but I'm a little bit confused by *how* a session gets authenticated. For basic auth or macaroons the logic is embedded in the `identity` body, but here we just seem to grab a `userid` from thin air. Is this helper using the (pre-validated?) session machinery internally?

Reply:
There's two APIs here:

1. ``unauthenticated_userid`` which is intended to pull the userid out of the request.
2. ``authenticated_userid`` which is intended to consume ``unauthenticated_userid`` and ensure that it matches what is in the data store.

For sessions, the data store is the session key, and when ``remember`` is called, the user id is stored in the user's Session, then ``authenticated_userid`` pulls that userid out of the session. Since session storage doesn't have passwords or anything like that and relies on the session id being randomly generated and secret, it just accepts whatever is in the session storage as the authenticated user id.

Reply:
Makes sense, thanks for the explanation. For some reason I forgot that session tokens are bearers and that that's the trust relationship.